### PR TITLE
Wait for Istio proxies after Reset

### PR DESCRIPTION
**Pull request by cnvergence**
See also #294

### CODE REVIEW

File: pkg/reconciler/instances/istio/reset/pod/handler.go [hunk 1]
```diff
@@ -45,8 +45,6 @@ type NoActionHandler struct {
 }
 
 func (i *NoActionHandler) Execute(object CustomObject, wg GetSyncWG) {
-	defer wg().Done()
```
Comment (mjakobczyk) [inline pkg/reconciler/instances/istio/reset/pod/handler.go:39]: After removing this `defer` please also remove unused `wg GetSyncWG` function parameter, regenerate mocks for this type and update tests.

File: pkg/reconciler/instances/istio/reset/pod/handler.go [hunk 2]
```diff
@@ -67,8 +65,6 @@ type DeleteObjectHandler struct {
 }
 
 func (i *DeleteObjectHandler) Execute(object CustomObject, wg GetSyncWG) {
-	defer wg().Done()
```
Comment (mjakobczyk) [inline pkg/reconciler/instances/istio/reset/pod/handler.go:51]: As above.

File: pkg/reconciler/instances/istio/reset/pod/handler.go [hunk 3]
```diff
@@ -103,8 +99,6 @@ type RolloutHandler struct {
 }
 
 func (i *RolloutHandler) Execute(object CustomObject, wg GetSyncWG) {
-	defer wg().Done()
```
Comment (mjakobczyk) [inline pkg/reconciler/instances/istio/reset/pod/handler.go:77]: As above.

**Review by mjakobczyk — CHANGES_REQUESTED** [review]

File: pkg/reconciler/instances/istio/reset/pod/handler.go [hunk 4]
```diff
@@ -140,3 +183,132 @@ func doRollout(customObject CustomObject, kubeClient kubernetes.Interface) (err
...
+		ready := isDeploymentReady(dep, i.kubeClient)
+		return ready, nil
+	})
+	if err != nil {
+		return err
+	}
+	i.log.Infof("%s/%s/%s is ready", object.Kind, object.Namespace, object.Name)
+
+	return nil
+}
```
Comment (piotrkpc) [inline pkg/reconciler/instances/istio/reset/pod/handler.go:263]: iiuc, `wait.Poll` seams to bring no new functionalities comparing to `retry` package that we are using in other cases. it also brought the `WaitOptions` without need imho. Would it be possible to use `retry` package here and not bringing another lib/pattern?

**Review by piotrkpc — CHANGES_REQUESTED** [review]
imo, code can be simplified by not bringing another dep for handling retries.

File: pkg/reconciler/instances/istio/actions/performer.go [hunk 5]
```diff
@@ -26,8 +27,10 @@ const (
 	istioOperatorKind     = "IstioOperator"
 	istioImagePrefix      = "istio/proxyv2"
 	retriesCount          = 5
-	delayBetweenRetries   = 10
+	delayBetweenRetries   = 5
 	sleepAfterPodDeletion = 10
+	timeout               = 5 * time.Minute
+	interval              = 10 * time.Second
 )
```
[the remark this answers is not in the document]
Comment (cnvergence) [reply pkg/reconciler/instances/istio/actions/performer.go:33]: Good Idea, I should have edited all the config at once
Also found not used config values.

File: pkg/reconciler/instances/istio/reset/pod/handler.go (hunk 4 above)
Comment (cnvergence) [reply pkg/reconciler/instances/istio/reset/pod/handler.go:263]: Does bring the functionality to poll with the condition function.
Retry is used here for reliability of k8s API calls, but lacks the simplicity of polling the state with condition func with strict timeout configuration, which I feel is much better for the task such like this one.

**Review by piotrkpc — APPROVED** [review]

**Review by mjakobczyk — APPROVED** [review]

File: pkg/reconciler/instances/istio/reset/pod/handler.go [hunk 6]
```diff
@@ -3,30 +3,49 @@ package pod
...
-type GetSyncWG func() *sync.WaitGroup
-
 //go:generate mockery --name=Handler --outpkg=mocks --case=underscore
-// Handler executes actions on the Kubernetes cluster
+// Handler executes actions on the Kubernetes objects.
 type Handler interface {
-	Execute(CustomObject, GetSyncWG)
+	// Execute action on the Kubernetes object with regards of the type of handler.
+	// Returns error if action was unsuccessful or wait timeout was reached.
+	ExecuteAndWaitFor(CustomObject) error
```
Comment (tobiscr) [inline pkg/reconciler/instances/istio/reset/pod/handler.go:23]: We have already an implememtation for such wait-actions. Please see example implementation here:

https://github.com/kyma-incubator/reconciler/blob/main/pkg/reconciler/kubernetes/progress/tracker_test.go#L110-L118

File: pkg/reconciler/instances/istio/reset/pod/handler.go [hunk 7]
```diff
@@ -35,75 +54,99 @@ type NoActionHandler struct {
...
-	Kind string
+	return i.WaitForResources(object)
 }
 
-// CustomObject contains all necessary fields to do rollout.
-type CustomObject struct {
-	Name      string
-	Namespace string
-	Kind      string
+func (i *RolloutHandler) WaitForResources(object CustomObject) error {
```
Comment (tobiscr) [inline pkg/reconciler/instances/istio/reset/pod/handler.go:122]: this is acutally the same thing the progress tracker is doing - please don't use your implementation for it.

Comment (tobiscr) [reply pkg/reconciler/instances/istio/reset/pod/handler.go:122]: @lindnerby : may I ask you to also have a look on the features provided by the new progress tracker to ensure we are not missing anything. THX a lot!

**Review by tobiscr — CHANGES_REQUESTED** [review]

File: pkg/reconciler/instances/istio/reset/pod/handler.go (hunk 6 above)
Comment (cnvergence) [reply pkg/reconciler/instances/istio/reset/pod/handler.go:23]: I guess both of them were created at the same time, I have added an issue about the follow-up.
https://github.com/kyma-incubator/reconciler/issues/369

**Comment by tobiscr** [comment]
Let's merge it with the existing code duplication and clean it up aftwards in #369

**Review by tobiscr — APPROVED** [review]

File: pkg/reconciler/instances/istio/reset/pod/handler.go (hunk 7 above)
Comment (lindnerby) [reply pkg/reconciler/instances/istio/reset/pod/handler.go:122]: Yes, there exists a `pkg/reconciler/kubernetes/progress/ready.go` which is providing functions for ready checks. You may have a look to reuse them here.

**Pull request merged**